demos/bbm_auxiliary/demo_bbm_aux.py.rst

```diff
@@ -0,0 +1,192 @@
+do quite well, but are inexact for the cubic one.  Here, we consider the
+reformulation in Andrews and Farrell, "Enforcing conservation laws and dissipation
+inequalities numerically via auxiliary variables" (arXiv:2407.11904, to appear
+in SIAM J. Scientific Computing) that preserves the third invariant at
+the expense of the second.  This method has an auxiliary variable in the system
+and requires a continuously differentiable spatial discretization (1D Hermite
+elements in this case).  The time discretization puts the main unknown in a
+continuous space and the auxiliary variable in a discontinuous one.  See
+equation (7.17) of Andrews' thesis for the particular formulation.
+
+
+Firedrake and Irksome imports::
+
+  from firedrake import (Constant, Function, FunctionSpace,
+      PeriodicIntervalMesh, SpatialCoordinate, TestFunction, TrialFunction,
+      assemble, derivative, dx, errornorm, exp, grad, inner,
+      interpolate, norm, project, replace, solve, split,
+  )
+
+  from irksome import Dt, GalerkinTimeStepper, TimeQuadratureLabel
+
+  def sech(x):
+      return 2 / (exp(x) + exp(-x))
+
+
+  N = 8000
+  L = 100
+  h = L / N
+  msh = PeriodicIntervalMesh(N, L)
+
+  c = Constant(0.5)
+
+  t = Constant(0)
+  dt = Constant(10*h)
+
+  x, = SpatialCoordinate(msh)
+
+  center = 40.0
+  delta = -c * center
+
+  uexact = 3 * c**2 / (1-c**2) \
+      * sech(0.5 * (c * x - c * t / (1 - c ** 2) + delta))**2
+
+  space_deg = 3
+  time_deg = 1
+
+This sets up the function space for the unknown :math:`u` and
+auxiliary variable :math:`\tilde{wH}`::
+
+  V = FunctionSpace(msh, "Hermite", space_deg)
+  Z = V * V
+
+We need a consistent initial condition for :math:`\tilde{wH}`. ::
+
+  def h1inner(u, v):
+      return inner(u, v) + inner(grad(u), grad(v))
+
+  def I1(u):
+      return u * dx
+
+  def I2(u):
+      return h1inner(u, u) * dx
+
+  def I3(u):
+      return (u**2 / 2 + u**3 / 6) * dx
+
+  uwHtilde = Function(Z)
+  uinit, wHinit = uwHtilde.subfunctions
+  
+  v = TestFunction(V)
+  w = TrialFunction(V)
+  a = h1inner(w, v) * dx
+  Finit = derivative(I3(uinit), uinit, v)
+
+  solve(a == h1inner(uexact, v)*dx, uinit)
+  solve(a == Finit, wHinit)
+
+Output the initial condition to disk::
+
+  xs = msh.coordinates.dat.data
+
+  with open("bbm_aux_init.csv", "w") as outfile:
+      outfile.write("x,u\n")
+      for xcur, ucur in zip(xs, uwHtilde.subfunctions[0].dat.data[::2]):
+          outfile.write("%f,%f\n" % (xcur, ucur))
+
+Create time quadrature labels::
+  
+  time_order_low = 2 * (time_deg - 1)
+  time_order_high = 3 * time_deg - 1
+
+  Llow = TimeQuadratureLabel(time_order_low)
+  Lhigh = TimeQuadratureLabel(time_order_high)
+
+This tags several of the terms with a low-order time integration scheme,
+but forces a higher-order method on the nonlinear term::
+
+  u, wHtilde = split(uwHtilde)
+  v, vH = split(TestFunction(Z))
+
+  lhs = h1inner(Dt(u) + wHtilde.dx(0), v) * dx + h1inner(wHtilde, vH) * dx
```

Comment:
This is simpler than the formulation in Boris' thesis since you've done IBP to lump the +/- terms together (which works due to the BCs)?